coalals/langserver.py

```diff
@@ -0,0 +1,148 @@
+from jsonrpc.endpoint import Endpoint
+from jsonrpc.dispatchers import MethodDispatcher
+from jsonrpc.streams import JsonRpcStreamWriter, JsonRpcStreamReader
+from .utils.files import UriUtils, FileProxyMap
+from .interface import coalaWrapper
+from .results import Diagnostics
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+class LangServer(MethodDispatcher):
+
+    def __init__(self, rx, tx):
+        self.root_path = None
+        self._dispatchers = []
+        self._shutdown = False
+
+        self._jsonrpc_stream_reader = JsonRpcStreamReader(rx)
+        self._jsonrpc_stream_writer = JsonRpcStreamWriter(tx)
+        self._endpoint = Endpoint(self, self._jsonrpc_stream_writer.write)
+        self._proxy_map = FileProxyMap()
+
+        # max_jobs is strict and a new job can only be submitted by pre-empting
+        # an older job or submitting later. No queuing is supported.
+        self._coala = coalaWrapper(self._proxy_map, max_jobs=2, max_workers=2)
+
+        self._capabilities = {
+            'capabilities': {
+                'textDocumentSync': 1,
+            }
+        }
+
+    def start(self):
+        self._jsonrpc_stream_reader.listen(self._endpoint.consume)
+
+    def local_p_analyse_file(self, filename, force=False):
+        """
+        Schedule a concurrent analysis if possible and
+        return a Future or a boolean depending on the
+        allocation.
+        """
+        proxy = self._proxy_map.resolve(filename, self.root_path)
```

Comment:
```coalaWrapper``` has access to ```_proxy_map```. It should be responsible to resolve the proxy itself.

Reply:
It should handle a proxy resolution failing because of failed attempt to open file.